### ISSUE TITLE
Refactor deepinv/optim

### ISSUE THREAD

**Issue by Tmodrzyk** [deepinv/deepinv#1147]
Let me first start by saying that I'm very grateful for all the optimization algorithms that are implemented in DeepInverse: I use this part of the library a lot for my research. The goal of this issue is not to complain but to try and improve the maintainability / extensibility of the library. I do not propose actionable solutions but rather try to diagnose the current issues. 

## TLDR
my feeling is that the optim module has become significantly harder to maintain and extend than other parts of the codebase. Several symptoms:
- the execution flow for even simple algorithms is very complex, going through several class inheritance and files
- the use of dictionaries makes the module harder to maintain
- there is a lot of redundancy within `optimizers.py`

In modules such as `physics`, newcomers have successfully implemented new forward operators related to their respective fields. The users should also be able to implement baseline algorithms to go along these operators.

-----------

To illustrate this, consider the example of the _Proximal Gradient Descent_ algorithm.
PGD is one of the simplest optimization algorithms we have:  
it alternates between a gradient step on the `DataFidelity` term and a proximal step on the `Prior`.  
A minimal implementation can be written quite naturally using `deepinv` components (`DataFidelity`, `Physics`, `Prior`), as shown for instance in [`demo_custom_optim.py`](https://deepinv.github.io/deepinv/auto_examples/basics/demo_custom_optim.html).

Now suppose a user wants to understand what happens when calling the built-in `PGD` algorithm.
They would roughly have to follow this path:

1. they first go to the PGD class and find that it does not implement a `forward` contrary to other models (DRUNet, RAM, TVDenoiser etc.)
2. it inherits from `BaseOptim` and basically instantiates it with some parameters (among them the most important one being: `iterator=PGDIteration` which dictates which iteration is called at each loop of the fixed point algorithm).
3. `BaseOptim` inherits from `Reconstructor` and implements a `forward`. Great ! The forward is essentially a call to `self.fixed_point`.
4. `self.fixed_point` is in turn initialized with the class `FixedPoint` with a bunch of parameters that are implemented within `BaseOptim` (`update_params_fn`, `update_prior_fn`, `update_data_fidelity_fn` etc.)
5. the user then goes to the implementation of the `FixedPoint` class to see what its forward is doing, it is located in another file `fixed_point.py`
6. `FixedPoint`'s forward is essentially a loop of calls to `self.single_iteration` which is implemented in this class
7. The ``single_iteration`` function does some calls to `FixedPoint`'s attribute functions (`update_params_fn`, `update_prior_fn`, `update_data_fidelity_fn` again) and one to `self.iterator`
8. In the init the user can see `self.iterator=iterator`
9. This param is handled to `FixedPoint` by `BaseOptim`, and it was provided in the init function of the `PGD` class.
10. The user is back in the init of `PGD`, where he sees that `iterator=PGDIteration`
11. The `PGDIteration` class is implemented in another file `pgd.py`. It does not implement a `forward`, however it inherits from `OptimIterator` which is implemented in another file `optim_iterator`. 
12. `OptimIterator` does implement a `forward` which calls several attribute functions `f_step`, `g_step`, `cost_fn` etc.
13. `f_step` is set to `fStepPGD` in `PGDIteration`'s init function, in `pgd.py`
14. `fStepPGD`'s implements a forward that actually does the expected gradient step, essentially:

```python
    return x - grad
```

Of course, each of these abstractions may make sense in isolation, but taken together it results on a very complicated flow. And this complexity is uncorrelated to the actual complexity of the optimization algorithm at hand: PGD is fairly simple, and the exact same flow would apply to plain _Gradient Descent_ anyway.

Note that I also simplified the story here, the same complexity emerges for every feature in the `BaseOptim` class: measuring convergence, momentum, backtracking, initialization...

----------------

Another difficulty I ran into is the way data is passed between components.  
A lot of the internal state is carried through a dictionary `X`.  
Depending on the algorithm, `X` can contain very different objects: current iterate, previous iterates, dual variables, cost values, hyperparameters, iteration count, etc.

The dictionary gives flexibility, but it also makes the code harder to reason about:
it is not always clear to know the expected structure of the dict for a given algorithm, nor what keys must be updated or not.

----------------

I know that some of the abstractions in the module help making the algorithms easily convertible into unrolled / DEQ schemes, and that `optim` has already been massively refactored in #592. However I think most of the complexity was already present before this refactor (most of the code is 2-3 years old).

---------------

Finally the `optimizers.py` file is very redundant: almost all of it is wrappers for the call of `BaseOptim`'s init. Most of the file is also very large docstrings, which makes it hard to go to the actual code. In other modules such as `physics`, these large docstrings are usually located at the very top of the file before the init, and there is only one class per file. 
This is way easier to work with.

---------------------

All the previous points are issues I faced when working on the module, but I also run GPT5.2 to see if it catches some stuff. Here is an automated audit which may point at bugs within the module:

- Undefined attribute in a “stop” path: fixed_point.py:350 prints self.max_iter_backtracking, but the file never defines it; this can turn “graceful stop after too many backtracking failures” into an AttributeError when verbose output is enabled.
- Unconditional debug output in a tight loop: data_fidelity.py:707 prints rel_crit every iteration, which is a strong side-effect (performance + noisy logs) and inconsistent with the rest of the module’s “verbose-gated” printing.
- TensorList support appears inconsistent in BiCGSTAB: bicgstab.py:61 uses torch.finfo(b.dtype) even though b can be a TensorList (the same function branches on TensorList earlier). Passing a TensorList b will likely fail here.
- Device/dtype hazards from literal tensors: default preprocessing uses torch.tensor(-5.0) without matching device/dtype, which can error on GPU tensors. See phase_retrieval.py:23 and spectral_methods.py:31.

**Comment by mh-nguyen712** [comment]
Thanks for making this diagnosis @Tmodrzyk! I do agree with you that it's hard to work with (and to maintain). 
I think we could discuss this seriously with our optim godfather cc @samuro95 😄

**Comment by jscanvic** [comment]
Thanks for the detailed report @Tmodrzyk I've rarely used that part of the library but I've heard the same feedback from a couple of people.

Here's my two cents. I agree that there's a lot of complexity going on and I do believe we can simplify things but we should remember that this complexity comes at the benefit of a very flexible framework. The hard question is how do we simplify things without loosing in flexibility?

One possible solution would be to have separate, simpler implementations of some of the algorithms (think actual for loops with explicit prox/grad and data fidelity steps). We could have automated tests to make sure they behave the same as implementations using the current framework. 

Of course it leaves the question of which algorithms to extract and re-implement. I fear some form of combinatorial explosion but it's worth investigating. It's also a bit unsatisfactory because it's kinda like duplicating code but I think it's fine in this case.

**Comment by Zeppo1994** [comment]
I am also strongly in favour of attempting a refactor, potentially at the Hackathon? When looking for how exactly parameters take effect, I had the same experience of clicking through an endless list of files...it has become easier to pin down with LLM usage now, but still its not exactly a great situation.

The current structure is also one of the reason why the nmAPG from our learned regularizers repository is still not merged. I would also have a LBFGS port that could be added with relatively low effort once this thing is cleaned up (it's experimental but should be well tested by then). Other opinions?

Cheers,
Sebastian

**Comment by Tmodrzyk** [comment]
Hi Sebastian, thanks for contributing to the discussion. 
I would be willing to attempt a refactor at the hackathon, but this is a massive piece of code, and at the heart of the library.
We attempted a refactor of the `Trainer` last year and it ended up being too much to handle, and the PR has gone stale. 
Any refactor should be carefully planed and sliced up into *very* small bits.

**Comment by samuro95** [comment]
Hi Sebastian,

Thanks for the comment!

The optimization refactoring is indeed one of the hackathon objectives. The current structure was designed around fixed-point iterations of optimization blocks, which is why FixedPoint is the base class. So there is no need to dig further :)

We can discuss during the hackathon how to make this structure more flexible.

Regarding the nmAPG PR, you are right that it is currently difficult to integrate the optimization algorithm you proposed into the existing structure. The plan is to incorporate it as part of the BaseOptim refactoring. Before that, we are currently working on integrating the prior-related part of the PR.

Best,
Samuel

**Comment by Tmodrzyk** [comment]
I like the solution @jscanvic proposed about keeping the old `deepinv/optim` around and adding our refactored alternative implementation, with tests to make sure that they are equivalent. 
This would enable us to progress on this without breaking everything, and taking ages to fix it before being able to merge.
